course_UvA-DL/activation-functions/Activation_Functions.py

```diff
@@ -0,0 +1,824 @@
+# In case the download below fails, you can download the models from a [Google Drive folder](https://drive.google.com/drive/folders/1sFpZUpDJVjiYEvIqISqfkFizfsTnPf4s?usp=sharing).
+# Please let me (Phillip) know if an error occurs so it can be fixed for all students.
+
+# %%
+# Github URL where saved models are stored for this tutorial
+base_url = "https://raw.githubusercontent.com/phlippe/saved_models/main/tutorial3/"
+# Files to download
+pretrained_files = [
+    "FashionMNIST_elu.config", "FashionMNIST_elu.tar", "FashionMNIST_leakyrelu.config",
+    "FashionMNIST_leakyrelu.tar", "FashionMNIST_relu.config", "FashionMNIST_relu.tar",
+    "FashionMNIST_sigmoid.config", "FashionMNIST_sigmoid.tar", "FashionMNIST_swish.config",
+    "FashionMNIST_swish.tar", "FashionMNIST_tanh.config", "FashionMNIST_tanh.tar"
+]
+# Create checkpoint path if it doesn't exist yet
+os.makedirs(CHECKPOINT_PATH, exist_ok=True)
+
+# For each file, check whether it already exists. If not, try downloading it.
+for file_name in pretrained_files:
+    file_path = os.path.join(CHECKPOINT_PATH, file_name)
+    if not os.path.isfile(file_path):
+        file_url = base_url + file_name
+        print(f"Downloading {file_url}...")
+        try:
+            urllib.request.urlretrieve(file_url, file_path)
+        except HTTPError as e:
+            print(
+                "Something went wrong. Please try to download the file from the GDrive folder, or contact the author with the full output including the following error:\n",
+                e
+            )
+
+# %% [markdown]
+# ## Common activation functions
+
+# %% [markdown]
+# As a first step, we will implement some common activation functions by ourselves.
+# Of course, most of them can also be found in the `torch.nn` package (see the [documentation](https://pytorch.org/docs/stable/nn.html#non-linear-activations-weighted-sum-nonlinearity) for an overview).
+# However, we'll write our own functions here for a better understanding and insights.
+#
+# For an easier time of comparing various activation functions, we start
+# with defining a base class from which all our future modules will
+# inherit:
+
+
+# %%
+class ActivationFunction(nn.Module):
+
+    def __init__(self):
+        super().__init__()
+        self.name = self.__class__.__name__
+        self.config = {"name": self.name}
+
+
+# %% [markdown]
+# Every activation function will be an `nn.Module` so that we can integrate them nicely in a network.
+# We will use the `config` dictionary to store adjustable parameters for some activation functions.
+#
+# Next, we implement two of the "oldest" activation functions that are still commonly used for various tasks: sigmoid and tanh.
+# Both the sigmoid and tanh activation can be also found as PyTorch functions (`torch.sigmoid`, `torch.tanh`) or as modules (`nn.Sigmoid`, `nn.Tanh`).
+# Here, we implement them by hand:
+
+# %%
+##############################
```

Comment:
is this needed? maybe not at least the first and last as it is already wrapped in a cell